Reduces friction for CUJ "I want to know how to use the Tabs API" (#3102)
GoogleChrome/developer.chrome.com: site/en/docs/extensions/reference/tabs/index.md

---
api: tabs
---

## Overview

The Tabs API not only offers features for manipulating and managing tabs, but can also detect the
[language][tabs-detect-language] of the tab, take a [screenshot][tabs-capture], and
[communicate][tabs-message] with a content script. 

{% Aside %}

The Tabs API can be used in the service worker and extension pages, but not in content scripts.

{% endAside %}

## Manifest {: #manifest }

Most functions don't require any permission, like, for example: [creating][tabs-create] a new tab,
[reloading][tabs-reload] a tab, [navigating][tabs-update] to another URL, etc.

### Permissions {: #perms }

> A permission is only required to access the `url`, `pendingUrl`, `title`, and `favIconUrl` of

Reviewer: I think it may be helpful to group these 4 properties together and refer to them collectively as "sensitive Tab properties" or something along those lines. This would give us a shorthand way of referring to them collectively and to discuss them in other contexts. For example, we would be able to say something like "host permissions grant an extension the ability to interact with sensitive properties on matching tabs."

Author: Good idea, I revised this sentence.

> [`tabs.Tab`][tab] or to capture a screenshot. Consider the following when requesting permissions:

Author: I rephrased this with your suggestion. Thanks!

> [Host permissions][doc-match] : With host permissions, the extension has access to the tab data of
> the URL match patterns and the ability to capture screenshots.

Reviewer: Took a shot at rewriting this to add a bit more detail. I didn't call out that script/style injection only applies to Manifest V2 here because a) I couldn't think of a way to do so without introducing confusion and b) the API docs for the relevant methods have deprecation warnings. WDYT?

```suggestion
[Host permissions][doc-match] : Host permissions grant an extension the ability to interact with
sensitive properties on matching tabs. They also allow an extension to perform more dangerous Tabs
API actions that require user consent such as [injecting content scripts](#method-executeScript),
[inserting](#method-insertCSS)/[removing](#method-removeCSS) CSS, and [capturing
screenshots](#method-captureVisibleTab) on matching pages.
```

> the URL match patterns and the ability to capture screenshots.

Author: It was a definition list originally! I think the rewrap moved it to the same line! My bad! I'll double-check next time 😅

uses `chrome.tabs.move`, you can use the same waiting pattern for other calls that modify tabs while
a drag may be in progress.

<web-tabs>
  <web-tab  title="Manifest V3 (promise)">

  ```js
  chrome.tabs.onActivated.addListener(activeInfo => move(activeInfo));

  async function moveToFirstPosition(activeInfo) {
    try {
      await chrome.tabs.move(activeInfo.tabId, {index: 0});
      console.log('Success.');
    } catch (error) {
      if (error == 'Error: Tabs cannot be edited right now (user may be dragging a tab).') {
        setTimeout(() => moveToFirstPosition(activeInfo), 50);
      }
    }
  }
  ```

{% Aside 'success' %}

The use of catch(error) in a Promise context is a way to ensure that an error that otherwise
populates chrome.runtime.lastError is not unchecked. 

{% endAside %}
  </web-tab>
  <web-tab title="Manifest V2 (callback)">

```js
chrome.tabs.onActivated.addListener((activeInfo) =>
  moveToFirstPositionMV2(activeInfo)
);

function moveToFirstPositionMV2(activeInfo) {
  chrome.tabs.move(activeInfo.tabId, { index: 0 }, () => {
    if (chrome.runtime.lastError) {
      const error = chrome.runtime.lastError;

>       if (
>         error ==
>         'Error: Tabs cannot be edited right now (user may be dragging a tab).'
>       ) {

Reviewer: Tweak the formatting of the `if` statement to better match the MV3 version.

```suggestion
      if (error == 'Error: Tabs cannot be edited right now (user may be dragging a tab).') {
```

---
api: tabs
---

## Overview

The Tabs API not only offers features for manipulating and managing tabs, but can also detect the
[language][tabs-detect-language] of the tab, take a [screenshot][tabs-capture], and
[communicate][tabs-message] with a tab's content scripts.

{% Aside %}

The Tabs API can be used by the service worker and extension pages, but not content scripts.

{% endAside %}

## Permissions {: #perms }

Most features do not require any permissions to use. For example: [creating][tabs-create] a new tab,
[reloading][tabs-reload] a tab, [navigating][tabs-update] to another URL, etc.

There are three permissions developers should be aware of when working with the Tabs API.

The "tabs" permission

: This permission does not give access to the `chrome.tabs` namespace. Instead, it
  grants an extension the ability to call [`tabs.query()`](#method-query) against four

>   sensitive properties on [`tabs.Tab`][tab] instances: `url`, `pendingUrl`, `title`, and

Author: In this case, [Tab](https://developer.chrome.com/docs/extensions/reference/tabs/#type-Tab) it is a type, and they are in [Pascal case](https://khalilstemmler.com/blogs/camel-case-snake-case-pascal-case/#Pascal-case-usage-in-TypeScript)

---
api: tabs
---

## Overview

The Tabs API not only offers features for manipulating and managing tabs, but can also detect the
[language][tabs-detect-language] of the tab, take a [screenshot][tabs-capture], and
[communicate][tabs-message] with a tab's content scripts.

{% Aside %}

The Tabs API can be used in the service worker and extension pages, but not in content scripts.

{% endAside %}

## Manifest {: #manifest }

Most functions don't require any permission, like, for example: [creating][tabs-create] a new tab,
[reloading][tabs-reload] a tab, [navigating][tabs-update] to another URL, etc.

### Permissions {: #perms }

> One of the following permissions is required to access sensitive [`tabs.Tab`][tab] properties, like
> the `url`, `pendingUrl`, `title`, and `favIconUrl` or to capture a screenshot:
> 
> [Host permissions][doc-match]
> 
> : <div class="stack">
> 
>   Host permissions grant access to sensitive Tab properties on matching tabs. They also allow an
>   extension to perform more intrusive actions that require user consent such as [injecting content
>   scripts](#method-executeScript), [inserting](#method-insertCSS)/[removing](#method-removeCSS) CSS,
>   and [capturing screenshots](#method-captureVisibleTab) on matching pages.
> 
>   {% Aside %}
> 
>   A better alternative for capturing the visible tab or current tab data, is the [`"activeTab"`][doc-activetab]
>   permission, which doesn't trigger any warnings.
> 
>   {% endAside %}
> 
>   </div>
> 
> Tabs permission
> 
> : The `"tabs"` permission allows access to sensitive Tab properties.

Reviewer: ```suggestion
Most features of the Tabs API do not require special permissions to use. For example, extensions can
call `chrome.tabs.create()` without declaring any permissions.

There are three permissions developers should be aware of when working with the tabs API.

The `tabs` permission

: This permission does not give the extension access to the `chrome.tabs` namespace. Instead, it
  grants an extension the ability to read and call [`tabs.query()`](#method-query) against four
  sensitive properties on [`tabs.Tab`][tab] instances: `url`, `pendingUrl`, `title`, and
  `favIconUrl`.

Host permissions

: [Host permissions][doc-match] allow an extension to read and query a matching tab's four sensitive
  `tabs.Tab` properties. They also grant an extension the ability to interact directly
  with a matching tabs using methods such as
  [`tabs.captureVisibleTab()`](#method-captureVisibleTab),
  [`tabs.executeScript()`](#method-executeScript), [`tabs.insertCSS()`](#method-insertCSS), and
  [`tabs.removeCSS()`](#method-removeCSS).

The `activeTab` permission

: [`activeTab`][doc-activetab] allows an extension to receive a temporary host permission grant in
  response to a user invocation. Unlike host permissions, `activeTab` does not trigger any warnings.
```

      // `tab` will either be a `tabs.Tab` instance or `undefined`.
      callback(tab);
    });
  }
  ```

  </web-tab>
</web-tabs>

### Mute the specified tab

This example shows how an extension can toggle the muted state for a given tab.

<web-tabs>
  <web-tab  title="Manifest V3 (promise)">

  ```js
  async function toggleMuteState(tabId) {
    const tab = await chrome.tabs.get(tabId);
    const muted = !tab.mutedInfo.muted;
    await chrome.tabs.update(tabId, {muted});
    console.log(`Tab ${tab.id} is ${muted ? 'muted' : 'unmuted'}`);
  }
  ```

</web-tab>
<web-tab  title="Manifest V2 (callback)">

  ```js
  function toggleMuteState(tabId) {
    chrome.tabs.get(tabId, (tab) => {
      let muted = !tab.mutedInfo.muted;
      chrome.tabs.update(tabId, {muted}, () => {
        console.log(`Tab ${tab.id} is ${muted ? 'muted' : 'unmuted'}`);
      });
    });
  }
  ```

  </web-tab>
</web-tabs>

### Move the current tab to the first position when clicked

This example shows how to move a tab while a drag may or may not be in progress. While this example
uses `chrome.tabs.move`, you can use the same waiting pattern for other calls that modify tabs while

> a drag may be in progress.

Reviewer 2: ```suggestion
a drag is in progress.
```

<web-tabs>
  <web-tab  title="Manifest V3 (promise)">

  ```js
  chrome.tabs.onActivated.addListener(moveToFirstPosition);

  async function moveToFirstPosition(activeInfo) {
    try {
      await chrome.tabs.move(activeInfo.tabId, {index: 0});
      console.log('Success.');
    } catch (error) {
      if (error == 'Error: Tabs cannot be edited right now (user may be dragging a tab).') {
        setTimeout(() => moveToFirstPosition(activeInfo), 50);
      } else {
        console.error(error);
      }
    }
  }
  ```

{% Aside 'success' %}

> The use of catch(error) in a Promise context is a way to ensure that an error that otherwise

Reviewer 2: ```suggestion
Using catch(error) in a Promise context is a way to ensure that an error that otherwise
```
